tests/helpers/diff_test_helpers.py

```diff
@@ -0,0 +1,147 @@
+        sys.exit("No fork before frontier")
+
+    forks = Hardfork.discover()
+
+    for i, fork in enumerate(forks):
+        if fork.short_name == fork_name:
+            return forks[i - 1].path[0], fork.path[0]  # type: ignore
+
+    return None, None
+
+
+def find_source_files(path: str) -> set:
+    """
+    Find .py files in a particular location
+    """
+    file_list = []
+    for directory, _, filenames in os.walk(path):
+        for filename in filenames:
+            if filename.endswith(".py"):
+                file_path = os.path.join(directory, filename)
+                file_path = file_path[len(path) + 1 :]
+                file_list.append(file_path)
+
+    return set(file_list)
+
+
+def process_source_code(path: str, name: str) -> None:
+    """
+    Clean up the source code.
+    Removes comments and empty lines.
+    """
+    if not os.path.isfile(path):
+        open(name, "a").close()
+        return
+
+    with open(path) as f:
+        code = f.read()
+
+    code = re.sub(re.compile('""".*?"""', re.DOTALL), "", code)
+    code = re.sub(re.compile("#.*?\n"), "\n", code)
+    code = re.sub(re.compile("\n\s*\n"), "\n", code)
+
+    with open(name, "w") as f:
+        f.write(code)
+
+
+class DiffTester:
+    """
+    Test the diffs in source code and compare it with expected diffs.
+    """
+
+    def __init__(
+        self,
+        fork_name: str,
+        expected_diffs_path: str,
+        found_diffs_path: str = None,
+    ) -> None:
+        self.old_fork_path, self.new_fork_path = get_fork_paths(fork_name)
+        self.expected_diffs_path = expected_diffs_path
+
+        if found_diffs_path:
+            self.found_diffs_path = found_diffs_path
+        else:
+            self.found_diffs_path = os.path.join(
+                os.path.dirname(os.path.realpath(__file__)), "found_diffs"
+            )
+        os.makedirs(self.found_diffs_path, exist_ok=True)
+
+    def create_diffs(self) -> None:
+        """
+        Create all the diffs of a fork wrt. the previous fork.
+        This will be the found diffs that can be compared with expected in `compare_diffs`
+        """
+        old_fork_files = find_source_files(self.old_fork_path)
+        new_fork_files = find_source_files(self.new_fork_path)
+
+        source_files = old_fork_files | new_fork_files
+
+        for file in source_files:
+            old_source_path = os.path.join(self.old_fork_path, file)
+            old_processed_path = os.path.join(self.found_diffs_path, "old.py")
+            new_source_path = os.path.join(self.new_fork_path, file)
+            new_processed_path = os.path.join(self.found_diffs_path, "new.py")
+
+            process_source_code(old_source_path, old_processed_path)
+            process_source_code(new_source_path, new_processed_path)
+
+            diffs = subprocess.run(
+                ["diff", old_processed_path, new_processed_path],
+                capture_output=True,
+                text=True,
+            )
```

Comment:
Hm... do you think it would be possible to implement a solution using Python's [`ast`](https://docs.python.org/3/library/ast.html)? You could walk the tree and remove docstrings and comments, plus whitespace would already be taken care of. Wouldn't need to shell out to `diff` that way either, or write files.

Apparently [there's no built-in comparison function](https://github.com/python/cpython/issues/60191) but [there's a post with some ideas on how to do it](https://stackoverflow.com/questions/3312989/elegant-way-to-test-python-asts-for-equality-not-reference-or-object-identity).

I guess the hardest part of this approach is specifying what changes are allowed. Could just hard code the test for `*_glacier` forks for now I guess.

Reply:
Sure. I will look into it.